local_vars/local_vars.cc

```diff
@@ -411,76 +420,82 @@ class LocalNameInserter {
+                newFun = core::Names::callWithSplatAndBlock();
+                newArgs.push_back(std::move(blockArg));
+                newNumPosArgs++;
             } else {
                 // <call-with-splat>(...)
-                original.fun = core::Names::callWithSplat();
+                newFun = core::Names::callWithSplat();
             }
         } else if (originalBlock == nullptr && shouldForwardBlockArg) {
-            // No positional splat and no "do", so we need to forward &<blkvar> with <call-with-block>.
-            original.reserveArguments(3 + posArgsEntries.size(), kwArgKeyEntries.size(),
-                                      /* hasKwSplat */ kwArgsHash != nullptr, /* hasBlock */ false);
-            original.addPosArg(std::move(original.recv));
-            original.addPosArg(std::move(method));
-            original.addPosArg(std::move(blockArg));
-
-            for (auto &arg : posArgsEntries) {
-                original.addPosArg(std::move(arg));
-            }
+            newArgs.reserve(3 + posArgsEntries.size() + kwArgKeyEntries.size() * 2 +
+                            /* hasKwSplat */ int(kwArgsHash != nullptr));
+            newArgs.push_back(std::move(newRecv));
+            newRecv = ast::MK::Magic(original.loc);
+            newArgs.push_back(std::move(method));
+            newArgs.push_back(std::move(blockArg));
+
+            absl::c_move(std::move(posArgsEntries), std::back_inserter(newArgs));
             posArgsEntries.clear();
 
+            newNumPosArgs = newArgs.size();
+
             if (kwArgsHash != nullptr) {
-                original.setKwSplat(std::move(kwArgsHash));
+                newArgs.push_back(std::move(kwArgsHash));
             } else {
                 ENFORCE(kwArgKeyEntries.size() == kwArgValueEntries.size());
                 for (size_t i = 0; i < kwArgKeyEntries.size(); i++) {
-                    original.addKwArg(std::move(kwArgKeyEntries[i]), std::move(kwArgValueEntries[i]));
+                    newArgs.push_back(std::move(kwArgKeyEntries[i]));
+                    newArgs.push_back(std::move(kwArgValueEntries[i]));
                 }
             }
             kwArgKeyEntries.clear();
             kwArgValueEntries.clear();
 
-            original.recv = ast::MK::Magic(original.loc);
-            original.fun = core::Names::callWithBlock();
+            newFun = core::Names::callWithBlock();
         } else {
             // No positional splat and we have a "do", so we can synthesize an ordinary send.
-            original.reserveArguments(posArgsEntries.size(), kwArgKeyEntries.size(),
-                                      /* hasKwSplat */ kwArgsHash != nullptr, /* hasBlock */ false);
+            newArgs.reserve(posArgsEntries.size() +
+                            kwArgKeyEntries.size() * 2
+                            /* hasKwSplat */
+                            + int(kwArgsHash != nullptr) + /* hasBlock */ int(originalBlock != nullptr));
```

Comment:
Nit: formats slightly better with `/* hasKwSplat */` after the `+`.
```suggestion
            newArgs.reserve(posArgsEntries.size() + kwArgKeyEntries.size() * 2 +
                            /* hasKwSplat */ int(kwArgsHash != nullptr) +
                            /* hasBlock */ int(originalBlock != nullptr));

```